### PR TITLE
Write references in compound datasets at the end

### LINKED ISSUE #144: [Bug]: Iterative write for compound reference datasets takes 5ever
### What happened?

hello hello! checking out how you're structuring the zarr export so I can mimic it, and first try wasn't able to complete an export because after ~10m or so i had only written ~6MB of ~3GB and so i got to profiling. Searched for an issue but didn't find one, sorry if this has been raised already.

The problem is here: 
https://github.com/hdmf-dev/hdmf-zarr/blob/9c3b4bee013c825e2cdaccbe443009f88302803e/src/hdmf_zarr/backend.py#L1021-L1026

where each row has to be separately serialized on write. 

I see this behavior elsewhere like in `__list_fill__` so it seems like this is a general problem with write indexing into object-type zarr arrays - when I tried to collect all the new items in a list and write them like `dset[...] = new_items`, i got a shape error (and couldn't fix it by setting the shape correctly in the `require_dataset` ).

What did work, though, is doing this:

```python
if len(refs) > 0:
    dset = parent.require_dataset(name,
                  shape=(len(data), len(data[0])),
                  dtype=object,
                  object_codec=self.__codec_cls(),
                  **options['io_settings'])
    self._written_builders.set_written(builder)  # record that the builder has been written
    dset.attrs['zarr_dtype'] = type_str
    
    new_items = []
    for j, item in enumerate(data):
        new_item = list(item)
        for i in refs:
            new_item[i] = self.__get_ref(item[i], export_source=export_source)
        new_items.append(new_item)

    dset[...] = np.array(new_items)
```

The only difference that I could see is that rather than returning a python list when accessing the data again, it returned an (object-typed) numpy array.

The difference in timing is considerable (in my case, tool-breaking, which is why i'm raising as a bug rather than a feature request). When converting [this dataset](https://api.dandiarchive.org/api/dandisets/000021/versions/draft/assets/4513d0c9-1e2b-4c8a-aa22-ae81822537c9/) with 9 compound datasets w/ references, putting a timer around the above block:

| Array write | Iterative Write |
| ----------- | -------------- |
| 0.03479409217834473 | 1.675 |
| 0.010181903839111328 | 0.205 |
| 0.19667983055114746 | 62.746 |
| 1.0213229656219482 | 1808.36 |
| 2.000258207321167 | (dnf, it's been 30m and i need to go home) |
| 0.31616997718811035 | ... |
| 0.0024509429931640625 | ... |
| 0.3125300407409668 | ... |
| 0.01106715202331543 | ... |

I only spent ~20m or so profiling and debugging, so it doesn't replicate current top-level behavior exactly (creates an object array rather than a list of lists), but imo it might be better: the result is the same shape as the source dataset (so, in this case, an n x 3 array) and can be indexed as such, where the current implementation can't index by column since the array is just a one-dimensional array of serialized python lists. It also looks like it would be possible to improve it further still by using a numpy structured dtype (zarr seems to be able to handle that, at least from the docs, but the `require_dataset` method doesn't like it), etc. but hopefully useful as a first pass.

The "steps to reproduce" below is just what I put in a test script and ran `cProfile` on to diagnose:

### Steps to Reproduce

```shell
from pynwb import NWBHDF5IO
from hdmf_zarr.nwb import NWBZarrIO

in_file = "/location/of/data/nwb/sub-738651046_ses-760693773.nwb"
out_file = "./test2.zarr"

with NWBHDF5IO(in_file, 'r', load_namespaces=True) as read_io:
    with NWBZarrIO(out_file, mode="w") as export_io:
        export_io.export(src_io=read_io, write_args={'link_data':False})
```

### Traceback

```shell
(not really a traceback kinda problem)
```

### Operating System

macOS

### Python Executable

Python

### Python Version

3.11

### Package Versions

[environment_for_issue.txt](https://github.com/hdmf-dev/hdmf-zarr/files/13596070/environment_for_issue.txt)

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/hdmf-dev/hdmf-zarr/blob/dev/.github/CODE_OF_CONDUCT.md)
- [X] Have you checked the [Contributing](https://github.com/hdmf-dev/hdmf-zarr/blob/dev/docs/CONTRIBUTING.rst) document?
- [X] Have you ensured this bug was not already [reported](https://github.com/hdmf-dev/hdmf-zarr/issues)?

### PR DESCRIPTION
**Pull request by mavaylon1**
## Motivation

What was the reasoning behind this change? Please explain the changes briefly.
Fix #144 

Notes:
The fix at face value is the following:
```
for j, item in enumerate(data):
        new_item = list(item)
        for i in refs:
             new_item[i] = self.__get_ref(item[i], export_source=export_source)
        new_items.append(tuple(new_item))
        
        generated_dtype = []
        for item in builder.dtype:
              if item['dtype'] == type('string'):
                  item['dtype'] = 'U25'
              generated_dtype.append((item['name'], item['dtype']))

 dset[...] = np.array(new_items, dtype=generated_dtype)
```

TLDR: We want to write outside of the the loop. What does this lead to?

To keep our compound dataset in its original form (i.e not expanded) we can use a structured array an pull the dtypes from the builder. However, it is required that the "data" in the array be in tuples and not lists (I believe it can also be in np.arrays but I have not tested this). 

Having these as tuples makes things a bit hard when retrieving data on read. For example, in the test `test_read_reference_compound`
```
def test_read_reference_compound(self):
        self.test_write_reference_compound()
        self.read()
        builder = self.createReferenceCompoundBuilder()['ref_dataset']
        read_builder = self.root['ref_dataset']
```
If I try to run `read_builder['data'][0]` I will get an error that `*** TypeError: 'tuple' object does not support item assignment`. Digging deeper it is because our `get_item` swaps/will reassign values in the tuple. This is not allowed because it is a tuple. Now I can convert that to a list on `_get_item_` and that does work. However, it is a little hacky. It also doesn't cover when my index is ":" vs a single integer. 

## How to test the behavior?
```
Show how to reproduce the new behavior (can be a bug fix or a new feature)
```

## Checklist

- [ ] Did you update CHANGELOG.md with your changes?
- [ ] Have you checked our [Contributing](https://github.com/hdmf-dev/hdmf-zarr/blob/dev/docs/CONTRIBUTING.rst) document?
- [ ] Have you ensured the PR clearly describes the problem and the solution?
- [ ] Is your contribution compliant with our coding style? This can be checked running `ruff` from the source directory.
- [ ] Have you checked to ensure that there aren't other open [Pull Requests](https://github.com/hdmf-dev/hdmf-zarr/pulls) for the same change?
- [ ] Have you included the relevant issue number using "Fix #XXX" notation where XXX is the issue number? By including "Fix #XXX" you allow GitHub to close issue #XXX when the PR is merged.

### CODE REVIEW

**Comment by mavaylon1** [comment]
@oruebel As I was going through the tests, I saw that the way I had to interact with the read data differed from before. Diving deeper it comes down to what I said above where we are now using tuples, which cannot change. The fix (even if it is a little hacky) is to convert to a list when trying to use index notation to "get". I wanted your thoughts on the matter and I also think this may need to go all the way up the parent hierarchy to HDMFDataset.

**Comment by oruebel** [comment]
> Digging deeper it is because our `get_item` swaps/will reassign values in the tuple.

Could you please point me to the lines of code where the tuple is being modified on read so I can take a look to better understand what is going. I'm not sure right now why the tuples are being modified on read (maybe to resolve the references).

**Comment by mavaylon1** [comment]
> > Digging deeper it is because our `get_item` swaps/will reassign values in the tuple.
> 
> Could you please point me to the lines of code where the tuple is being modified on read so I can take a look to better understand what is going. I'm not sure right now why the tuples are being modified on read (maybe to resolve the references).

[Lines ](https://github.com/hdmf-dev/hdmf-zarr/blob/9692a98f0568e0f968c89411a67bd4f483c91d9d/src/hdmf_zarr/zarr_utils.py#L153-L165)

File: src/hdmf_zarr/zarr_utils.py [hunk 1]
```diff
@@ -151,7 +151,8 @@ def dtype(self):
         return self.__dtype
 
     def __getitem__(self, arg):
-        rows = copy(super().__getitem__(arg))
+        rows = list(copy(super().__getitem__(arg)))
+        # breakpoint()
```
Comment (oruebel) [inline src/hdmf_zarr/zarr_utils.py:155]: If I understand what you are saying correctly, the issue appears when ``self.__swap_refs`` is being called because the individual rows are now represented as tuples. I.e, `rows` is either a tuple (if `arg` is an int) or a list of tuples (if `arg` selects multiple rows). If that is the issue, then I think the we could update ``self.__swap_refs`` to return a new instance of the row (rather than updating the existing one). Here is what I think might work: 


```
def __getitem__(self, arg):
        rows = copy(super().__getitem__(arg))
        if np.issubdtype(type(arg), np.integer):
            rows =  self.__swap_refs(rows)
        else:
            rows = [self.__swap_refs(row) for row in rows]
       return rows

def __swap_refs(self, row):
        updated_row = list(row)  # convert tuple to a list so we can update it
        for i in self.__refgetters:
            getref = self.__refgetters[i]
            updated_row[i] = getref(row[i])
        return updated_row   # TODO if we want to keep these as tuples then we could convert them back here 
```

**Comment by sneakers-the-rat** [comment]
I was just going to work on https://github.com/hdmf-dev/hdmf-zarr/pull/146 and saw this parallel PR, should I work on this instead or what

File: src/hdmf_zarr/backend.py [hunk 2]
```diff
@@ -1018,11 +1018,20 @@ def write_dataset(self, **kwargs):  # noqa: C901
...
                     for i in refs:
                         new_item[i] = self.__get_ref(item[i], export_source=export_source)
-                    dset[j] = new_item
+                    new_items.append(tuple(new_item))
+                    # dset[j] = new_item
+
+                generated_dtype = []
+                for item in builder.dtype:
+                    if item['dtype'] == type('string'):
+                        item['dtype'] = 'U25'
```
Comment (sneakers-the-rat) [inline src/hdmf_zarr/backend.py]: this mutates the builder's dtype right?

**Comment by sneakers-the-rat** [comment]
Wrote this here: https://github.com/hdmf-dev/hdmf-zarr/pull/146#issuecomment-1855049073

but that tuple problem is just from not setting the dtype in the zarr dataset, so it ignores the dtype you give to `np.array` and just stores as tuples. If you correctly set that then it behaves just fine.

So do this:
https://github.com/hdmf-dev/hdmf-zarr/blob/8100341512324af25718ab93dc7fc54b68f03469/src/hdmf_zarr/backend.py#L1044

instead of this:
https://github.com/hdmf-dev/hdmf-zarr/blob/005bfbc0eb3ecee884170ec2a93b7d20c63c9655/src/hdmf_zarr/backend.py#L1016

File: src/hdmf_zarr/backend.py (hunk 2 above)
Comment (mavaylon1) [reply src/hdmf_zarr/backend.py]: Yes. When numpy gets the "str class" it removes the actual value and on read returns an empty string.

**Comment by mavaylon1** [comment]
> I was just going to work on #146 and saw this parallel PR, should I work on this instead or what

Yes please

**Comment by mavaylon1** [comment]
> read_builder['data'][0]

Interesting. Could you explain more?

**Comment by mavaylon1** [comment]
> Wrote this here: [#146 (comment)](https://github.com/hdmf-dev/hdmf-zarr/pull/146#issuecomment-1855049073)
> 
> but that tuple problem is just from not setting the dtype in the zarr dataset, so it ignores the dtype you give to `np.array` and just stores as tuples. If you correctly set that then it behaves just fine.
> 
> So do this:
> 
> https://github.com/hdmf-dev/hdmf-zarr/blob/8100341512324af25718ab93dc7fc54b68f03469/src/hdmf_zarr/backend.py#L1044
> 
> instead of this:
> 
> https://github.com/hdmf-dev/hdmf-zarr/blob/005bfbc0eb3ecee884170ec2a93b7d20c63c9655/src/hdmf_zarr/backend.py#L1016

Oh it stores it as <class 'numpy.void'> when you provide a dtype instead of a tuple.

File: src/hdmf_zarr/backend.py (hunk 2 above)
Comment (sneakers-the-rat) [reply src/hdmf_zarr/backend.py]: not sure what you mean, when numpy gets `str` where? What i mean is that this would mutate the `builder.dtype` dict, which is almost always undesirable, especially since the builder is passed in as an arg, mutating your args will buy you invisible bugs later.

eg.

```python
>>> mydtypes = [{'dtype': 'int', 'name': 'test'}, {'dtype': 'float', 'name': 'test2'}]
>>> for item in mydtypes:
...    if item['dtype'] == 'int':
...        item['dtype'] = 'newvalue'
>>> mydtypes
[{'dtype': 'newvalue', 'name': 'test'}, {'dtype': 'float', 'name': 'test2'}]
```

instead do 

```python
for item in builder.dtype:
    if item['dtype'] is str:
        generated_dtype.append((item['name'], 'U25'))
    else:
        generated_dtype.append((item['name'], item['dtype']))
```

which doesn't mutate `builder.dtype`. see https://github.com/hdmf-dev/hdmf-zarr/blob/8100341512324af25718ab93dc7fc54b68f03469/src/hdmf_zarr/backend.py#L1028-L1036

**Comment by sneakers-the-rat** [comment]
> Yes please

Alright, lmk how - i don't have commit permissions on this repo so can't modify this branch. Usually bad etiquette to take the several hours of work i've done here and just rewrite it without credit, but whatevs.

> Oh it stores it as <class 'numpy.void'> when you provide a dtype instead of a tuple.

right, and that's the desired behavior for storing a structured array, which is what I thought the goal was. As this PR is written, the constructed `dtype` is just ignored and the array is just stored as serialized tuples.

File: src/hdmf_zarr/backend.py (hunk 2 above)
Comment (mavaylon1) [reply src/hdmf_zarr/backend.py]: I see what you did.

**Comment by mavaylon1** [comment]
> > Yes please
> 
> 
> 
> Alright, lmk how - i don't have commit permissions on this repo so can't modify this branch. Usually bad etiquette to take the several hours of work i've done here and just rewrite it without credit, but whatevs.
> 
> 
> 
> > Oh it stores it as <class 'numpy.void'> when you provide a dtype instead of a tuple.
> 
> 
> 
> right, and that's the desired behavior for storing a structured array, which is what I thought the goal was. As this PR is written, the constructed `dtype` is just ignored and the array is just stored as serialized tuples.

I believe we were working on this in parallel. We can just keep it to your branch on your fork to give you credit. The intent was for me to dig around from the last checkpoint I saw on yours. Let me know when you're done and I'll review it. I'll continue to play around here for educational purposes.

**Comment by sneakers-the-rat** [comment]
ya sorry i don't mean to be prickly but this has happened a number of times to me lol and usually i don't care, but when last looking for a job it actually was important to be able to demonstrate contribs. I think the patch is done except for tests and docs?

**Comment by mavaylon1** [comment]
> ya sorry i don't mean to be prickly but this has happened a number of times to me lol and usually i don't care, but when last looking for a job it actually was important to be able to demonstrate contribs. I think the patch is done except for tests and docs?

I can check the tests, but I will be out of town starting tomorrow. I'll go through it will most likely finalize this for a merge by Tuesday.

**Comment by sneakers-the-rat** [comment]
lmk what else needs to be written re: docs and tests, happy to do that! i try to be as labor-neutral as i can be when raising issues/PRs :). take ur time, no rush. also sorry for my tone earlier, frustrating day.

**Comment by rly** [comment]
@mavaylon1 can this be closed, now that #146 has been merged?
